data/src/main/java/com/linkedin/data/Data.java

```diff
@@ -372,18 +353,27 @@ private static void traverse(Object obj, TraverseCallback callback,
         }
         else
         {
-          checkForCyclesAndAdd(map, ancestorSet, pathList);
-          callback.startMap(map);
-          Iterable<Map.Entry<String, Object>> orderedEntrySet = callback.orderMap(map);
-          for (Map.Entry<String, Object> entry : orderedEntrySet)
+          try
+          {
+            if (map._isTraversing)
+            {
+              throw new IOException("Cycle detected.");
+            }
+
+            map._isTraversing = true;
```

Comment:
We have to be careful with this change. This would fail if a DataMap is traversed in multiple threads concurrently. I'm sure there are apps that do concurrent traversals. This might even happen within the context of a single request also.
Also, this might happen more often in tests sharing a common test input and run in parallel. 

I was thinking we should keep the existing algo, and provide a config for services to disable the cycle check (or use a count based check)

Reply:
I updated the existing algo to use ThreadLocal to get around your concern. It will still be faster than using an identityset.
I also added an API to customize the cycle checker.